src/server/pps/server/api_server.go

```diff
@@ -2955,6 +2959,46 @@ func (a *apiServer) deletePipelineInTransaction(txnCtx *txncontext.TransactionCo
 	return nil
 }
 
+// DeletePipelines implements the protobuf pps.DeletePipelines RPC.  It deletes
+// multiple pipelines at once.  If all is set, then all pipelines in all
+// projects are deleted; otherwise the requested pipelines are deleted.
+func (a *apiServer) DeletePipelines(ctx context.Context, request *pps.DeletePipelinesRequest) (response *pps.DeletePipelinesResponse, retErr error) {
+	var (
+		projects = make(map[string]bool)
+		dr       = &pps.DeletePipelineRequest{
+			Force:    request.Force,
+			KeepRepo: request.KeepRepo,
+		}
+		pp []*pps.Pipeline
+	)
+	for _, p := range request.GetProjects() {
+		projects[p.String()] = true
+	}
+	dr.Pipeline = &pps.Pipeline{}
+	pipelineInfo := &pps.PipelineInfo{}
+	deleted := make(map[string]struct{})
+	if err := a.pipelines.ReadOnly(ctx).List(pipelineInfo, col.DefaultOptions(), func(string) error {
+		if _, ok := deleted[pipelineInfo.Pipeline.String()]; ok {
+			// while the delete pipeline call will delete historical versions,
+			// they could still show up in the list.  Ignore them.
+			return nil
+		}
+		if !request.GetAll() && !projects[pipelineInfo.GetPipeline().GetProject().GetName()] {
+			return nil
+		}
+		dr.Pipeline = pipelineInfo.Pipeline
+		err := a.deletePipeline(ctx, dr)
+		if err == nil {
+			deleted[pipelineInfo.Pipeline.String()] = struct{}{}
+			pp = append(pp, pipelineInfo.Pipeline)
+		}
+		return err
```

Comment:
Maybe aggregate this error into a `multierror` as well?

Reply:
Ditto here — I think that it is probably better to terminate early rather than continue.